### PR TITLE
Basic OCSP support

### PR DESCRIPTION
**Pull request by NightBlues**
Hi! I've implemented basic OCSP support and seems to be working:)
My use-case: implement ocsp responder and configure nginx with `ssl_ocsp` and `ssl_ocsp_responder` parameters to check **client** certs without CRL lists.
~I'm going to add few unittests (interaction with openssl cli) before merge.~
added tests for interaction with openssl

### CODE REVIEW

**Comment by hannesm** [comment]
Thanks for your contribution. This looks great. The only concern I have is that the public interface (x509.mli) is extended with everything available in OCSP -- is this intentional? In other modules I expose only those types and functions that are needed by clients. Also, docstrings are missing (you can generate the API documentation locally by running `dune build @doc`, and see its output).

**Comment by NightBlues** [comment]
> In other modules I expose only those types and functions that are needed by clients. 

If i hide for example some types from `X509.OCSP.Request` library client will not be able to destructure it to obtain for example certID. Though i can add `make` function for creating `Request.t` with various content.
I think that i can hide `Asn` submodules - library clients only need few functions from them - decode and encode.

**Comment by hannesm** [comment]
@NightBlues it sounds good to have a `make` (or `create`) function. And yes, the `Asn` submodule shouldn't be needed. Is your client code of this PR open source?

**Comment by NightBlues** [comment]
> Is your client code of this PR open source?

No, but its very simple - just look in postgres for serialNumber to learn if it has been revoked and return response. I can create unittest emulating this. 
But rfc covers all these extensions that may be required for something:)

**Comment by NightBlues** [comment]
@hannesm I've reworked OCSP module interface, so its easier to use, but some features are now not available. Added unit test that emulates simple OCSP responder behavior.

**Comment by hannesm** [comment]
Hi @NightBlues, I reviewed your changes, and pushed a commit that adapts minor things: avoid camelCase in the API, cleanup some whitespaces, sign operation supported EC keys (by using x509_default_scheme instead of hardcoding `RSA_PKCS1). let me know if these changes are fine with you.

**Comment by hannesm** [comment]
one minor thing: in `Request.Asn.tbs_request` you explicitly set the (optional) `version` to `Some version`, while in `Response.Asn.response_data` you set the optional version to `None` if it is `version_v1` -- should we unify this to either always include the version_v1 or to never include it (or are there interoperability dragons hidden?)?
Related, should that version, if present, be verified in the decoder to be version_v1 (and result in a parse error otherwise), and not be represented in the higher level (see private_key.ml where the version is dropped already in the decode & encode)? I think failing early for unsupported versions is a good idea, and since t

**Comment by hannesm** [comment]
what I have in mind is 2a53cfd which (a) ensures request and response version is v1 if provided (b) does not emit a version in either request or response. once a version 2 is standardised and used, we can revise.

**Comment by NightBlues** [comment]
@hannesm thank you for help! Changes are good to me. I used `RSA_PKCS1 and `SHA256 because i saw it in openssl's output, i'm just afraid that some old version of openssl could not handle our modern algorithms:) Also openssl used to omit version in requests and responses, so its good idea to do the same.

**Comment by hannesm** [comment]
ok, well, the default for rsa keys is RSA_PKCS1, and for hash SHA256 (but recent enhancements allow EC keys and appropriate hashes depending on curves).

about the responderid:
```
-        (* there must not be [explicit 1], but openssl do so:( *)
-        (required ~label:"responderID" @@ explicit 1 @@ responder_id)
+        (required ~label:"responderID" responder_id)
```

so the asn grammar (and asn1-combinators) are slightly tricky in this case, let me rephrase from rfc6069:
```
   ResponseData ::= SEQUENCE {
      version              [0] EXPLICIT Version DEFAULT v1,
      responderID              ResponderID,
      producedAt               GeneralizedTime,
      responses                SEQUENCE OF SingleResponse,
      responseExtensions   [1] EXPLICIT Extensions OPTIONAL }

   ResponderID ::= CHOICE {
      byName               [1] Name,
      byKey                [2] KeyHash }
```

My intuition is that in ResponseData the first element (version) is optional, and the second a choice -- thus the second must be tagged (otherwise the decoder wouldn't know whether it is the first or second element. The index of this specific choice is starting from 1 (since 0 is the tag used for the first element (version)).

Thus, the way to encode it is to make a `choice2 (explicit 1 ..) (explicit 2 ...)` -- since a choice can be implicit (the default, works if all items have disjoint encodings). I did so in 86ab35c -- the tests are passing, is this fine with you? :)

```
    let responder_id : responder_id Asn.t =
      let f = function
        | `C1 dn -> `ByName dn
        | `C2 hash -> `ByKey hash
      in
      let g = function
        | `ByName dn -> `C1 dn
        | `ByKey hash -> `C2 hash
      in
      map f g @@
      choice2 (explicit 1 Distinguished_name.Asn.name) (explicit 2 octet_string)
```

**Comment by NightBlues** [comment]
> choice2 (explicit 1 ..) (explicit 2 ...)

Thats a good point! But there is still no `EXPLICIT` keyword inside ResponderID in rfc, so we follow openssl behavior instead of rfc for compatibility:) If tests pass my code should work too!

**Comment by NightBlues** [comment]
@hannesm Thank you for so active help with this feature!

**Comment by hannesm** [comment]
> Thats a good point! But there is still no EXPLICIT keyword inside ResponderID in rfc

This is true, but whether something is explicit or implicit tagged in a asn1 grammar is rather complicated; TL;DR: in this case with implicit tagging it could be ambiguous (if a ResponderID of type int would be allowed). And the choice in this location must be expicitly tagged.

The difference between your code (expicit 1 of choice) and mine (choice of explicit) is that yours will fail on ByKey (since only tag 1 is accepted in your grammar -- the encoder of yours is also wrong for ByKey).

**Comment by NightBlues** [comment]
> The difference between your code (expicit 1 of choice) and mine (choice of explicit) is that yours will fail on ByKey (since only tag 1 is accepted in your grammar -- the encoder of yours is also wrong for ByKey).

yes, that's good improvement!

**Comment by hannesm** [comment]
Thanks for your work. I merged this as is, though would like to add code that verifies a OCSP response (and eventually the request as well if there is a signature present) -- so a potential client can be implemented on top. I'll develop this code this weekend, and cut a X509 release.
